### PR TITLE
Endpoint for Gateway to fetch the fastest route

### LINKED ISSUE #6: [Pathfinding] Create endpoint to take in rooms and sensors and returns the fastest path

### PR DESCRIPTION
**Pull request by Mafusn**
This pull request includes significant changes to enhance the pathfinding functionality, improve code modularity, and update documentation. The most important changes include adding new methods for creating instances from schemas, refactoring routes, and adding comprehensive tests for the new pathfinding feature.

### Enhancements to Pathfinding Functionality:

* [`app/classes/room.py`](diffhunk://#diff-781a99729ac6aeb78e3ebcf5c734d0720055e1c07aebcc5dbc8e05feec254225R15-R37): Added `from_schema` and `create_room_mapping_from_schemas` methods to create Room instances from schemas.
* [`app/classes/sensor.py`](diffhunk://#diff-98906d07c09411ab47b33866ee6a26f6e20d0e030db2bc738e8313910182f3daR14-R46): Added `from_schema` and `create_sensors_from_schemas` methods to create Sensor instances from schemas and attach Rooms to them.
* [`app/services/route_service.py`](diffhunk://#diff-9ffd528cd70d30b71956419632d1a832d85b99ace7ed3b78ff3166aece8ba12aR1-R48): Implemented `create_fastest_path` function to process pathfinding requests and compute the fastest path using a sensor graph.

### Refactoring Routes:

* [`app/routes/__init__.py`](diffhunk://#diff-7ba09577cfb2a1dc76a02099ab4a0edbf94374b64d6cd5e74162b60c4196b73fR1-R7): Introduced a list of routers to modularize route inclusion.
* [`app/routes/pathfinding.py`](diffhunk://#diff-d3c932713b3ec343a360c5878488c63766516b9b1cfe13145702ebe3be46c528R1-R33): Added a new route to handle pathfinding requests and return the fastest path.
* [`app/main.py`](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddL3-R7): Refactored route imports and included routers dynamically. [[1]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddL3-R7) [[2]](diffhunk://#diff-990f7e4140fab1ad82afc787505090b64d4024e63a891405cd9085e7e6b249ddR20-R22)

### Adding Tests:

* [`app/test/routes/test_pathfinding.py`](diffhunk://#diff-8ba5255aaa97fdeaab98e2272a69f807d6cece34bcc2721d4f0e53e6b365ad01R1-R89): Added tests to verify pathfinding functionality, including scenarios for missing data, incorrect sensors, and successful pathfinding.
* [`app/test/services/test_route_service.py`](diffhunk://#diff-ad968d4e4d34d6a133a6f95b0f890e434cf7fc5deefeb07c26882abc18919482R1-R51): Added tests to check the `create_fastest_path` function for various edge cases and successful execution.

### Updating Type Annotations:

* [`app/classes/room.py`](diffhunk://#diff-781a99729ac6aeb78e3ebcf5c734d0720055e1c07aebcc5dbc8e05feec254225L2-R2): Updated `__init__` method to include type annotations for parameters.
* [`app/classes/sensor.py`](diffhunk://#diff-98906d07c09411ab47b33866ee6a26f6e20d0e030db2bc738e8313910182f3daL2-R2): Updated `__init__` method to include type annotations for parameters.

### Documentation and Configuration:

* [`.vscode/settings.json`](diffhunk://#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357R1-R7): Configured VSCode settings for Python testing using pytest.
* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L72-R72): Updated development and production commands, and added notes on debugging and error handling. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L72-R72) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R82-R90)

### CODE REVIEW

File: .vscode/settings.json
Comment (Jatewo) [inline .vscode/settings.json:1]: Is this supposed to be tracked in git?

File: README.md [hunk 1]
```diff
@@ -69,7 +69,7 @@ nix develop
 
 ### 🔹 **For Development**
 ```bash
-uvicorn app:app --reload
+uvicorn app.main:app --reload --port 8001
```
Comment (Jatewo) [inline README.md]: Can we define the por in main instead?

File: README.md [hunk 2]
```diff
@@ -79,3 +79,12 @@ gunicorn -k uvicorn.workers.UvicornWorker -w 4 -b 0.0.0.0:8080 app.main:app
 
 ---
 
+## Testing
+### Debug tests
+- Outcomment the entire "addopts" line in "root/pytest.ini"
```
Comment (Jatewo) [inline README.md:88]: I don't understand the purpose of this.

File: app/classes/room.py [hunk 3]
```diff
@@ -1,14 +1,29 @@
...
         self.crowd_factor = crowd_factor
 
     @classmethod
     def from_dict(cls, data: dict):
-        return cls(
-            id=data.get("id"),
-            name=data.get("name"),
-            crowd_factor=data.get("crowd_factor")
-        )
+        return cls(id=data.get('id'), name=data.get('name'), crowd_factor=data.get('crowd_factor'))
```
Comment (Jatewo) [inline app/classes/room.py]: Would typing like this:

```py
from typing import TypedDict

class dict_type(TypedDict):
	id: str
	name: str
	crowd_factor: float
```
 Make sense to add? I usually try to use it for more control and better type hints

Comment (Jatewo) [reply app/classes/room.py]: I guess this goes for all places where an unknown `dict` type annotation is given

File: app/main.py [hunk 4]
```diff
@@ -62,18 +66,16 @@ def main():
...
+                room_path.append(', '.join(common_rooms) if common_rooms else 'No common room')
 
-            print("Fastest path (room names along each transition):")
+            print('Fastest path (room names along each transition):')
             print(room_path)
-            print("Total effective weight (sum of crowd factors):", total_weight)
+            print('Total effective weight (sum of crowd factors):', total_weight)
         else:
-            print("No path found between the selected sensors.")
+            print('No path found between the selected sensors.')
```
Comment (Jatewo) [inline app/main.py]: Can we make this a guard clause instead?

File: app/main.py [hunk 5]
```diff
@@ -62,18 +66,16 @@ def main():
...
     else:
-        print("Not enough sensors to determine a path.")
+        print('Not enough sensors to determine a path.')
```
Comment (Jatewo) [inline app/main.py]: Same as above

File: app/routes/pathfinding.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+from ..services.route_service import create_fastest_path
+from ..schemas.path import FastestPathRequest
+
+router = APIRouter(prefix='/pathfinding')
+
+
+@router.post('/fastest-path')
+async def get_fastest_path(request_body: FastestPathRequest):
+    """
+    Accepts a JSON payload that contains optional lists of Rooms and Sensors,
```
Comment (Jatewo) [inline app/routes/pathfinding.py]: Can we use the swagger documentation fields instead of this long description? I can help if needed

File: ruff.toml [hunk 7]
```diff
@@ -2,5 +2,5 @@ line-length = 100
 exclude = ["app/not_used"]
 [format]
 quote-style = "single"
-indent-style = "tab"
+indent-style = "space"
```
Comment (Jatewo) [inline ruff.toml]: I think this is a mistake as well. We did settle on tab, as far as I know. Has this changed?

**Review by Jatewo — CHANGES_REQUESTED** [review]
Most of it looks good. I do, however, have some comments to address.

I also have some general issues:

1. The type annotations are generally lacking and gives quite a few issues with e.g. pylance
2. There are some parts of the functionality not currently being tests. I think we should at lease consider if we should test this.

File: .vscode/settings.json
Comment (Mafusn) [reply .vscode/settings.json:1]: It's nice to have the same settings if other people use VScode: [Source](https://stackoverflow.com/questions/32964920/should-i-commit-the-vscode-folder-to-source-control)

**Review by JackSMunn — APPROVED** [review]
Approving, but the comments from @Jatewo needs to be adressed before merging.

**Review by MortenMunk — APPROVED** [review]
lgtm

File: app/classes/room.py (hunk 3 above)
Comment (Mafusn) [reply app/classes/room.py]: This will not be implemented in this PR, but i have made an issue for it #8
